CIP-0102? | Royalty Datum Metadata (#551)
cardano-foundation/CIPs: CIP-XXXX/README.md

---
CIP: ?
Title: Royalty Datum Metadata

> Authors: Sam Delaney <sam@ikigaitech.org>

Reviewer: ```suggestion
Authors:
  - Sam Delaney <sam@ikigaitech.org>
```
`Authors` is a list in the new format.

> Comments-URI:

Reviewer: ```suggestion
Discussions:
```
This has changed name (and is also a list, if anything added in the future e.g. Cardano Forum threads).

Status: Proposed

> Type: Informational

Reviewer: ```suggestion
Category: Tokens
```
There is no longer a `Type` field; with a new `Category` field set as follows: https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#categories

Created: 2023-08-08

> Post-History:

Reviewer: ```suggestion
```
This field no longer exists.

---

> CIP: ????

Reviewer: ```suggestion
CIP: ?
```

Reviewer: (as specified in CIP-0001)

Author: Done. FYI, this syntax appears to cause an error message in the rendered version:
![image](https://github.com/cardano-foundation/CIPs/assets/20246757/201d3a7b-1cdf-4529-9496-9a8bf9c75741)

It shows up in other PRs as well: https://github.com/serejke/CIPs/blob/CIP-keccak256/CIP-%3F%3F%3F%3F/README.md

> Title: Royalty Datum Metadata Standard

Reviewer: ```suggestion
Title: Royalty Datum Metadata
```
I wanted to include @KtorZ @Ryun1 @SebastienGllmt in a discussion of whether we should use the term Standard here.  I would anticipate problems if so, because another metadata solution might also address royalties in the future and then we would have uncertainty over which one is "standard" while leaving that word out would force the unambiguous use of a merged CIP number.

(CIP27 wasn't so much a problem in this way because they used the commercial designation "CNFT" in addition to "Standard" in the merged CIP title.)

@SamDelaney since the term "standard" is used so many places in the document I would recommend just changing it in the header metadata & the PR title and leaving everything else as-is: to keep the tone & clarity of what you have already written, therefore leaving more detailed readers to make up their own mind over whether to consider this a "standard".

```

#### Example of onchain variable fee calculation:
```
; Given a royalty fee of 1.6% (0.016)

; To store this in the royalty datum
1 / (0.016 / 10) => 625

; To read it back
10 / 625 => 0.016
```
Because the computational complexity of Plutus primitives scales with size, this approach significantly minimizes resource consumption.

To prevent abuse, it is **recommended** that the `royalty NFT` is stored at the script address of a validator that ensures the specified fees are not arbitrarily changed, such as an always-fails validator.

### Reference Datum Royalty Flag

If not specified elsewhere in the token's datums, a malicious user could send transactions to a protocol which do not reference the royalty datum. For full assurances, a new optional flag should be added to the reference datum

```
extra = 
	{
		...

		? royalty_included : big_int
	}
```

- If the field is present and > 1 the validators must require a royalty input.
- If the field is present and set to 0 the validators don't need to search for a royalty input.
- If the field is not present, validators should accept a royalty input, but not require one.

### Examples

#### Retrieve metadata as 3rd party

> A third party has the following NFT `d5e6bf0500378d4f0da4e8dde6becec7621cd8cbf5cbb9b87013d4cc.(222)TestToken` and they want to lookup the royalties. The steps are

Reviewer: ```suggestion
A third party has the following NFT `d5e6bf0500378d4f0da4e8dde6becec7621cd8cbf5cbb9b87013d4cc.(222)TestToken` and they want to look up the royalties. The steps are
```
(this is spelled correctly other places in the document... `lookup` being a noun and `look up` being a verb phrase)

Author: Good catch thanks! Fixed.

1. Construct `royalty NFT` from `user token`: `d5e6bf0500378d4f0da4e8dde6becec7621cd8cbf5cbb9b87013d4cc.(500)Royalty`
2. Look up `royalty NFT` and find the output it's locked in.

> 3. Get the datum from the output and lookup metadata by going into the first field of constructor 0.

Reviewer: ```suggestion
3. Get the datum from the output and look up metadata by going into the first field of constructor 0.
```

### Acceptance Criteria

- [ ] This CIP should receive feedback, criticism, and refinement from: CIP Editors and the community of people involved with NFT projects to review any weaknesses or areas of improvement.
- [ ] Guidelines and examples of publication of data as well as discovery and validation should be included as part of of criteria for acceptance.
- [ ] Minimal reference implementation making use of [Lucid](https://github.com/spacebudz/lucid) (off-chain), [PlutusTx](https://github.com/input-output-hk/plutus) (on-chain): [Reference Implementation](./ref_impl/)
- [ ] Implementation and use demonstrated by the community: NFT Projects, Blockchain Explorers, Wallets, Marketplaces.

### Implementation Plan

1. Publish open source reference implementation and instructions related to the creation, storage and reading of royalty utxos.
2. Implement in open source libraries and tooling such as [Lucid](https://github.com/spacebudz/lucid), [Blockfrost](https://github.com/blockfrost/blockfrost-backend-ryo), etc.
3. Achieve additional "buy in" from existing community actors and implementors such as: blockchain explorers, token marketplaces, minting platforms, wallets.

> ## References

Reviewer: The CIP-0001 format, and therefore any parsers that may work on CIP material, has no _References_ section.  I would recommend putting these links inline: with the first use of a CIP term linked to its document & further references unlinked (like Wikipedia).

There is a format where anchors can be collected at the bottom of the page but I can't find a reference to it or a name for it... but some CIPs use it (e.g. CIP-0094) and you could model your linking on them.  Defining the links in that way would collect them at the bottom of the page which would help serve the purpose of your list here.

Author: I wasn't aware of that syntax, very cool. Updated as recommended.

> - CIP-0027: https://github.com/cardano-foundation/CIPs/blob/master/CIP-0027

Reviewer: ```suggestion
- CIP-0027: https://github.com/cardano-foundation/CIPs/tree/master/CIP-0027
```
The pathname component `/tree` is canonical (as already used for the others).

Author: References section has been removed per your comment below. The new anchored link is updated to use `/tree`.
